contracts/okp4-cognitarium/src/querier/plan_builder.rs

```diff
@@ -105,10 +105,43 @@ impl<'a> PlanBuilder<'a> {
     }
 
     fn build_triple_pattern(&mut self, pattern: &TriplePattern) -> StdResult<QueryNode> {
-        Ok(QueryNode::TriplePattern {
-            subject: self.build_subject_pattern(pattern.subject.clone())?,
-            predicate: self.build_predicate_pattern(pattern.predicate.clone())?,
-            object: self.build_object_pattern(pattern.object.clone())?,
+        let subject_res = self.build_subject_pattern(pattern.subject.clone());
+        let predicate_res = self.build_predicate_pattern(pattern.predicate.clone());
+        let object_res = self.build_object_pattern(pattern.object.clone());
+
+        let mut bound_variables: Vec<usize> = vec![];
+        let maybe_subject = match subject_res {
```

Comment:
Just a quick comment on the duplication of code for processing the 3 elements (subject, predicate, object). I guess that by introducing a generic local `resolve` function, we could reduce repetition and making the logic much clearer and more concise.

For instance (to be tested):

```rust
fn resolve<T>(
            component_res: StdResult<PatternValue<T>>,
            bound_variables: &mut Vec<usize>,
        ) -> Result<Option<PatternValue<T>>, StdError> {
            Ok(match component_res {
                Ok(value) => {
                    value.lookup_bound_variable(&mut |v| bound_variables.push(v));
                    Some(value)
                }
                Err(err) if NamespaceResolver::is_ns_not_found_error(&err) => None,
                Err(err) => Err(err),
            })
        }
```

And then:

```rust
let mut bound_variables: Vec<usize> = vec![];

let maybe_subject = resolve(subject_res, &mut bound_variables)?;
let maybe_predicate = resolve(predicate_res, &mut bound_variables)?;
let maybe_object = resolve(object_res, &mut bound_variables)?;
```

Just a suggestion. I'll leave it to you to decide what's best.